Add seekable method to Tech and increase supported version range

stevendesu opened the pull request:
I initially just needed to resolve #41, but I figured I'd go ahead and tweak the `package.json` while I was at it. Sorry for the confusing triple-commit (you can squash-merge it if you prefer). Since we were forking the repo I made a few changes for convenience for our sake and reverted them prior to the pull request.

jthomerson commented:
Also, you need to run `grunt standards && npm test` to find errors there. You'll see here that the build failed (https://travis-ci.org/silvermine/videojs-chromecast/jobs/498762630#L501) because of a linting issue.

stevendesu commented:
`git reset origin/master` doesn't do anything (since there are no changes in my working tree), but I can squash the three commits into one which should clean it up. I know the process, but can never remember the specific commands - so I'll have to Google them quickly.

jthomerson commented:
@stevendesu I just realized why that won't work for you ... you forked the repo and used the `master` branch, so _your_ `origin/master` is _your_ repo. So, reset won't work there because you're resetting to HEAD.

Instead, you could add the official origin and reset to it, e.g.

```
git remote add silvermine https://github.com/silvermine/videojs-chromecast.git
git fetch silvermine
git reset silvermine/master
```

Or, you could have used a feature branch (the recommended flow) instead of master.

Anyway, thanks for working with us. Just thought maybe that explanation would be helpful for the future.

stevendesu commented:
@jthomerson Using a squash commit I cleaned up the history for you. It may not be the most elegant solution (feature branches would probably be smarter), but there's only a single commit now.

As for `grunt standards`, there is an issue being thrown by eslint and I may need some direction for the preferred method of solving it. To return a `TimeRange` in the `seekable` method I needed access to the VideoJS API. Because this is stored on the global `window` object I was able to just type `videojs.createTimeRange()` and it worked when I tested it, but `videojs` is ***technically*** not declared in this scope, nor is it passed in as a parameter anywhere.

Does VideoJS make its API available to all components (perhaps through `this.videojs`) or is there a preferred way to pass it in?

stevendesu commented:
Sure enough, `this.videojs` is a thing :smile: I just inspected the player in the debugger and checked it out. Since it worked I replaced `videojs.createTimeRange` with `this.videojs.createTimeRange` then range:

```bash
grunt standards && npm test
```

No errors seemed to be thrown. One new rebase and force-push and we're looking good.

jthomerson commented:
@stevendesu thanks. Not sure if https://github.com/silvermine/videojs-chromecast/blob/52c3b840f5457d28cf2a29e068e7c8e42fa8d04f/src/js/tech/ChromecastTech.js#L38 provides what you need?

@yokuze can you look at the questions @stevendesu asks about the exposure of `videojs`?

jthomerson commented:
@yokuze cancel that.

jthomerson commented:
But, assigning the rest of the (actual code) review to @yokuze

yokuze commented:
@jthomerson sure thing. Hi @stevendesu, thanks again for the PR! Taking a look now.

yokuze commented on src/js/tech/ChromecastTech.js:
```diff
@@ -364,6 +366,30 @@ ChromecastTech = {
+    * VOD. Perhaps if there's a way to detect if the source is live then we could adjust
+    * the seekable behavior accordingly.
+    *
+    * @returns {undefined} always returns `undefined`
```
This function is returning a `TimeRanges` instance. Could you update this `@returns` statement to reflect that?

Also, could you note in the description of the `@returns` statement that the `TimeRanges` object that's returned always contains only one `TimeRange` object, and that `TimeRange` is always `0` through the `duration` of the current media item? Something like:

```
    * @returns {TimeRanges} always returns a `TimeRanges` object with one `TimeRange` that
    * starts at `0` and ends at the `duration` of the current media item
```

yokuze commented on src/js/tech/ChromecastTech.js:
```diff
@@ -364,6 +366,30 @@ ChromecastTech = {
+    * how much the media item has playable content, so we'll just assume the entire video
+    * is an available seek target.
+    *
+    * The risk here lies in DVR, where ther may exist a sliding window and seeking is only
```
For readers that are not familiar with the `DVR` acronym, could we either:

1) Link to a resource (preferably some official documentation somewhere) that explains what `DVR` is, or:
2) Use a more familiar, generic term like `live streaming`

yokuze replied:
Small typo: `where ther` -> `where there`.

Also, it may be slightly more clear to say `sliding window of playable content` rather than just `sliding window`.

yokuze commented on src/js/tech/ChromecastTech.js:
```diff
@@ -364,6 +366,30 @@ ChromecastTech = {
+    *
+    * Unfortunately we have no way of detecting when this is the case. Returning anything
+    * other than the full range of the video means that we lose the ability to seek during
+    * VOD. Perhaps if there's a way to detect if the source is live then we could adjust
```
The last sentence seems like it's more applicable to developers who work on this plugin, rather than developers that _use_ the plugin. Could you move this last sentence to a TODO comment in the body of this function, and instead end the paragraph with something like:

```
    * VOD. Therefore, we just return a `TimeRanges` object with a single `TimeRange` that
    * starts at `0` and ends at the `duration` of the current media item.
```

then:

```
      // TODO Investigate if there's a way to detect if the source is live, so that we can
      // possibly adjust the seekable `TimeRanges` accordingly.
      return this.videojs.createTimeRange(0, this.duration());
```

Thanks!

yokuze commented on src/js/tech/ChromecastTech.js:
```diff
@@ -163,6 +163,8 @@ ChromecastTech = {
       this.trigger('waiting');
       this._clearSessionTimeout();
 
+      mediaInfo.duration = this.duration();
```
This sets the duration correctly for the first item that's played, but if the `src` changes (e.g. if someone is playing items one right after another in a playlist of some sort), `this.duration()` will return the `duration` of the previous `src` and `mediaInfo.duration` will be incorrect.

Is this even strictly necessary? It would be nice to be able to set this property, but the Chromecast seems to be able to calculate the duration and play fine without us setting it. We don't know what the `duration` of the `src` is at this point in the case I described above.

yokuze commented:
@stevendesu just finished the initial review. Thanks!

stevendesu commented:
@yokuze Thank you for the thorough comments. I've made all of the requested changes and updated the PR. I also added one additional "Refs" to the commit message: #32 

Reading through the other issues, this one mentioned that the reason seeking failed was the lack of `seekable` method. Now that a `seekable` method is implemented, seeking works on VOD.

yokuze commented on src/js/tech/ChromecastTech.js:
```diff
@@ -364,6 +364,33 @@ ChromecastTech = {
+    * other than the full range of the video means that we lose the ability to seek during
+    * VOD.
+    *
+    * @returns {TimeRanges} always returns a `TimeRnages` object with one `TimeRnage` that
```
A couple of typos here: `TimeRnages` and `TimeRnage`

stevendesu replied:
Damn, I make a lot of those 😮

stevendesu commented:
@yokuze Two typos corrected (I noticed since GitHub is a single-page app now, I had to refresh the page to see the changes take effect in the "Files changed" tab) and commit message updated

jthomerson commented:
Thanks @stevendesu. I just released `@silvermine/videojs-chromecast@1.1.2` with your fix.
